src/librustc/traits/error_reporting.rs

```diff
@@ -523,6 +524,11 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
     {
         let span = obligation.cause.span;
 
+        if !self.reported_selection_errors.borrow_mut().insert((span, error.clone())) {
```

Comment:
Yeah, it was surprising to me how many tests were affected by this. I'll think of a different way to deduplicate these errors...

Originally I was trying to get to the point where only one of `FnMut` and `FnOnce` were required, which feels like the correct way to accomplish this, but could not find where to tackle that (even pouring through the code and with full debugging output on :-/ ).